File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int zeroCount = 0;
        int[] res = new int[nums.length];
        int product = 1;

        for (int n : nums) {
            if (n == 0) zeroCount++;
            else product *= n;
        }

        if (zeroCount > 1) return res;

        for (int i = 0; i < nums.length; i++) {
            if (zeroCount == 1)
                res[i] = (nums[i] == 0) ? product : 0;
            else res[i] = product / nums[i];
        }

        return res;
    }
}  

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        cur_min, cur_max = 1, 1

        for n in nums:
            tmp = cur_max * n
            cur_max = max(cur_max * n, cur_min * n, n)
            cur_min = min(cur_min * n, tmp, n)
            res = max(cur_max, res)

        return res